hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCleaner.java

```diff
@@ -158,8 +159,7 @@ public HoodieCleaner(HoodieTable<T> hoodieTable, HoodieWriteConfig config) {
 
         // determine if we have enough commits, to start cleaning.
         if (commitTimeline.countInstants() > commitsRetained) {
-            HoodieInstant earliestCommitToRetain =
-                commitTimeline.nthInstant(commitTimeline.countInstants() - commitsRetained).get();
+            HoodieInstant earliestCommitToRetain = getEarliestCommitToRetain().get();
             List<List<HoodieDataFile>> fileVersions =
                 fileSystemView.getEveryVersionInPartition(partitionPath)
```

Comment:
@prazanna 

For this particular case, a partition had 129758 file patterns to delete, while other partitions had < 100. So Collections.shuffle did not help. Why this partition is receiving so many updates is yet another issue we need to find out, but seems like it is.

globStatus in HoodieCleaner.clean() was the bottleneck (log excerpt below of the hung job). So, I parallelized the globStatus + delete operations. I ran the change on a copy of the table that had clean failing for a week, and the cleaning worked perfectly fine (with a repartition by 200 all the delete tasks finished almost at same time). Yesterday, I isolated and deployed it to the production table since its situation had gotten worse (num file patterns to delete for the skewed partition had doubled since I ran the fix on its copy a week back) and the clean successfully completed.

1. the globStatus in HoodieCleaner.clean() was the bottleneck and not getEveryVersionInPartition(). Since I wanted to re-use most existing code, ended up making partition level calls to fetch file patterns to delete.

2. yes, globStatus is the problem, but I have also noticed that deletes can sometimes take > 30-50s, in which case, the skew avoidance would still be required to avoid clean completion delays.

@vinothchandar my observation - doing this based on commit metadata would definitely help a lot if listStatus calls are lessened. We would still need to parallelize deletes because of 2).

```
java.io.IOException: Failed on local exception: java.io.InterruptedIOException: Interruped while waiting for IO on channel java.nio.channels.SocketChannel. 59986 millis timeout left.
..
	at org.apache.hadoop.fs.FileSystem.globStatus(FileSystem.java:1625)
	at com.uber.hoodie.io.HoodieCleaner.clean(HoodieCleaner.java:238)
```